base/iobuffer.jl

```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
+    # If reinit, the data has been handed out to the user, and the IOBuffer
+    # no longer controls it, so we need to allocate a new one.
+    if !io.writable || io.reinit
+        return ensureroom_reallocate(io, nshort)
+    end
+    # The fast path here usually checks there is already room, then does nothing.
+    # When append is true, new data is added after io.size, not io.ptr
+    existing_space = min(lastindex(io.data), io.maxsize) - (io.append ? io.size : io.ptr - 1)
+    if existing_space < nshort % Int
+        # Outline this function to make it more likely that ensureroom inlines itself
+        return ensureroom_slowpath(io, nshort)
     end
-    copyto!(io.data, 1, io.data, ptr, bytes_to_move)
-    io.size -= ptr - 1
-    io.ptr -= ptr - 1
+    return io
+end
+
+# Throw error (placed in this function to outline it) or reinit the buffer
+@noinline function ensureroom_reallocate(io::GenericIOBuffer, nshort::UInt)
+    io.writable || throw(ArgumentError("ensureroom failed, IOBuffer is not writeable"))
+    io.data = _similar_data(io, min(io.maxsize, nshort % Int))
+    io.reinit = false
     io.offset = 0
-    return
+    return io
 end
 
 @noinline function ensureroom_slowpath(io::GenericIOBuffer, nshort::UInt)
-    io.writable || throw(ArgumentError("ensureroom failed, IOBuffer is not writeable"))
-    if io.reinit
-        io.data = _similar_data(io, nshort % Int)
-        io.reinit = false
+    # Begin by zeroing out offset and check if that gives us room enough
+    if !iszero(io.offset)
+        nshort_after_zero_offset = (nshort % Int) - zero_offset!(io)
+        nshort_after_zero_offset < 1 && return io
+        nshort = nshort_after_zero_offset % UInt
     end
-    if !io.seekable
-        if !ismarked(io) && io.ptr > io.offset+1 && io.size <= io.ptr - 1
-            io.ptr = 1
-            io.size = 0
-            io.offset = 0
-        else
-            datastart = (ismarked(io) ? io.mark : io.ptr - io.offset)
-            if (io.size-io.offset+nshort > io.maxsize) ||
-                (datastart > 4096 && datastart > io.size - io.ptr) ||
-                (datastart > 262144)
-                # apply somewhat arbitrary heuristics to decide when to destroy
-                # old, read data to make more room for new data
-                compact(io)
-            end
+
+    data_len = min(lastindex(io.data), io.maxsize)
+
+    # Else, try to compact the data. To do this, the buffer must not be seekable.
+    # If it's seekable, the user can recover used data by seeking before ptr,
+    # and so we can't delete it.
+    if (!io.seekable && io.ptr > 1)
+        ptr = io.ptr
+        mark = io.mark
+        size = io.size
+        data = io.data
+        to_delete = (mark > -1 ? min(mark, ptr - 1) : ptr - 1)
+        # Only shift data if:
+        if (
+                # It will prevent us from having to resize buffer, or
+                to_delete >= nshort % Int ||
```

Comment:
I realize this is sort of what we did before (with a threshold value of 4096 instead of 1 though), but this is not actually a good idea to have here since it can make writing to a PipeBuffer `O(n^2)` in the worst case. We should instead resize anytime that we would recover less than 32 bytes or less than 1/8th of the buffer, whichever is larger

Reply:
Did as you suggested, except I also added a check to shift in case `io.maxsize` is low enough that resizing wouldn't free enough space.